kazoo/recipe/lock.py

```diff
@@ -242,18 +253,34 @@ def _inner_acquire(self, blocking, timeout):
             finally:
                 self.client.remove_listener(self._watch_session)
 
-    def acquired_lock(self, children, index):
-        return index == 0
+    def predecessor(self, children, index):
+        for c in children[:index]:
+            if any(n in c for n in self._EXCLUDE_NAMES):
```

Comment:
The predecessor function returns the first node ahead of us in
the list which is mutually exclusive with the kind of lock we are
trying to acquire.  So read it as "names which are exclusive with
us" rather than "names to exclude from the list".

In the case of a write lock, the exclude list includes "rlock"
and "lock", so this will return the first of either of those
types -- meaning that if there is a client ahead of us with
either a read or write lock, we have not acquired the lock and
must wait for that client to release it (they are a predecessor).

In the case of a read lock, the exclude list includes
only "lock", so this will return the first write lock entry
only (our predecessor in that case).  It will ignore any read
lock entries ahead of us.

So I believe the logic is correct.  I will note, however, that
the current code will set the watch on the last node which
preceded our node.  This is efficient because if there are N
predecessors, then we would only set the watch on the latest and
would therefore only wake up and go through the loop after all
the predecessors had released their locks.  This patch, however,
sets the watch on the first node which precedes ours, meaning
that we will wake up and iterate through the loop for all N
predecessor nodes.  I'll push up a patch to correct that.